compiler/rustc_codegen_llvm/src/builder/autodiff.rs

```diff
@@ -0,0 +1,363 @@
+    //  %0 = fmul double %x, %x
+    //  ret double %0
+    // }
+    // ```
+    //
+    // The user now applies autodiff to the function square, in which case fn_to_diff will be `square`.
+    // Our macro generates the following placeholder code (slightly simplified):
+    //
+    // ```llvm
+    // define double @dsquare(double %x) {
+    //  ; placeholder code
+    //  return 0.0;
+    // }
+    // ```
+    //
+    // so our `outer_fn` will be `dsquare`. The unsafe code section below now removes the placeholder
+    // code and inserts an autodiff call. We also add a declaration for the __enzyme_autodiff call.
+    // Again, the arguments to all functions are slightly simplified.
+    // ```llvm
+    // declare double @__enzyme_autodiff_square(...)
+    //
+    // define double @dsquare(double %x) {
+    // entry:
+    //   %0 = tail call double (...) @__enzyme_autodiff_square(double (double)* nonnull @square, double %x)
+    //   ret double %0
+    // }
+    // ```
+    unsafe {
+        // On LLVM-IR, we can luckily declare __enzyme_ functions without specifying the input
+        // arguments. We do however need to declare them with their correct return type.
+        // We already figured the correct return type out in our frontend, when generating the outer_fn,
+        // so we can now just go ahead and use that. FIXME(ZuseZ4): This doesn't handle sret yet.
+        let fn_ty = llvm::LLVMGlobalGetValueType(outer_fn);
+        let ret_ty = llvm::LLVMGetReturnType(fn_ty);
+
+        // LLVM can figure out the input types on it's own, so we take a shortcut here.
+        let enzyme_ty = llvm::LLVMFunctionType(ret_ty, ptr::null(), 0, True);
+
+        //FIXME(ZuseZ4): the CC/Addr/Vis values are best effort guesses, we should look at tests and
+        // think a bit more about what should go here.
+        let cc = llvm::LLVMGetFunctionCallConv(outer_fn);
+        let ad_fn = declare_raw_fn(
+            cx,
+            &ad_name,
+            llvm::CallConv::try_from(cc).expect("invalid callconv"),
+            llvm::UnnamedAddr::No,
+            llvm::Visibility::Default,
+            enzyme_ty,
+        );
+
+        // Otherwise LLVM might inline our temporary code before the enzyme pass has a chance to
+        // do it's work.
+        let attr = llvm::AttributeKind::NoInline.create_attr(llcx);
+        attributes::apply_to_llfn(ad_fn, Function, &[attr]);
+
+        // first, remove all calls from fnc
+        let entry = llvm::LLVMGetFirstBasicBlock(outer_fn);
+        let br = llvm::LLVMRustGetTerminator(entry);
+        llvm::LLVMRustEraseInstFromParent(br);
+
+        let last_inst = llvm::LLVMRustGetLastInstruction(entry).unwrap();
+        let mut builder = Builder::build(cx, entry);
+
+        let num_args = llvm::LLVMCountParams(&fn_to_diff);
+        let mut args = Vec::with_capacity(num_args as usize + 1);
+        args.push(fn_to_diff);
+
+        let enzyme_const = cx.typeid_metadata("enzyme_const".to_string()).unwrap();
```

Comment:
This should probably get a separate method -- this is metadata, but not typeid metadata (which is a CFI thing).

Reply:
`cx.typeid_metadata` is just a thin wrapper around `LLVMMDStringInContext2` and does nothing else, which I used to call directly, before moving to use cx: CodegenCx. https://github.com/rust-lang/rust/blob/904d8f6b39ce98fb890ac3381a710f071ddae44d/compiler/rustc_codegen_llvm/src/type_.rs#L310

I guess it just got introduced along with cfi functions and I can just adjust the wrapper name?

Reply:
I also checked the original IR created by my full fork and it doesn't include any of the typeid strings there: https://llvm.org/docs/TypeMetadata.html#testing-addresses-for-type-membership
(otherwise enzyme also wouldn't pick them up, since we match the name from the beginning)

Reply:
> `cx.typeid_metadata` is just a thin wrapper around `LLVMMDStringInContext2` and does nothing else, which I used to call directly, before moving to use cx: CodegenCx.
> 
> https://github.com/rust-lang/rust/blob/904d8f6b39ce98fb890ac3381a710f071ddae44d/compiler/rustc_codegen_llvm/src/type_.rs#L310
> 
> I guess it just got introduced along with cfi functions and I can just adjust the wrapper name?

Yes, the method internally does the right thing, the API is just not intended for this. It's part of the TypeMembershipCodegenMethods trait. Places that generate non-typeid metadata seem to currently just call LLVMMDStringInContext2 directly, but could also add a separate wrapper.

Reply:
I added a new wrapper, create_metadata.